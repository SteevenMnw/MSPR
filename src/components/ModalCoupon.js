//import liraries
import React, { useState, useEffect } from "react";
import { View, Text, StyleSheet, Modal, Pressable } from "react-native";

import { getCouponById, addCouponForUser } from "../API/API_Access";

// create a component
const ModalCoupon = (props) => {
  const { visible, idCoupon } = props;
  const [stateModal, setStateModal] = useState(true);
  const [coupon, setCoupon] = useState({});
  const [playOnce, setPlayOnce] = useState(false);
  const [couponError, setCouponError] = useState(false);
  const [user, setUser] = useState([]);

  useEffect(() => {
    if (!isNaN(idCoupon)) {
      getCouponById(idCoupon)
        .then((response) => {
          setCoupon(response);
          setPlayOnce(true);
        })
        .catch(() => {
          noDisplayModal("Aucun coupon est associé à ce QR Code !");
        });
    } else {
      noDisplayModal("Aucun coupon est associé à ce QR Code !");
    }

    if (coupon.info == 0) {
      noDisplayModal("Le coupon que vous avez scanné n'est plus disponible !");
    }
  }, [playOnce]);

  function noDisplayModal(textAlertError) {
    setStateModal(false);
    setCouponError(true);
    alert(textAlertError);
  }

  function getDifferenceInDays() {
    const date1 = new Date();
    const date2 = new Date(coupon.date_end);
    const time_diff = date2.getTime() - date1.getTime();

    const days_diff = (time_diff / (1000 * 3600 * 24)).toFixed(0);

    return days_diff;
  }

  getUserSession = async () => {
    try {
      const value = await AsyncStorage.getItem("user");
      setUser(JSON.parse(value));
    } catch (e) {
      console.log(e);
    }
  };

  function addCoupon() {
    addCouponForUser(user.id_user, coupon.id_coupon)
      .then(() => {
        alert("Coupon ajouté à votre liste !");
        setStateModal(false);
      })
      .catch(() => alert("Coupon déjà enregistré à votre liste !"));
  }

  return (
    <Modal
      animationType="fade"
      transparent={true}
      visible={stateModal ? visible : stateModal}
      onRequestClose={() => {
        setStateModal(false);
      }}
    >
      {!couponError && (
        <View style={styles.centeredView}>
          <View style={styles.modalView}>
            <View>
              <Text style={styles.titleText}>Description du coupon : </Text>
              <Text style={styles.modalText}>{coupon.description}</Text>
              {coupon.date_end && (
                <View>
                  <Text style={styles.titleText}>Date d'expiration : </Text>
                  <Text style={styles.modalText}>
                    Il reste {getDifferenceInDays()} jour(s) pour l'utiliser.
                  </Text>
                </View>
              )}
              {coupon.compteur && (
                <View>
                  <Text style={styles.titleText}>Utilisations : </Text>
                  <Text style={styles.modalText}>
                    Il reste {coupon.compteur} utilisations de ce coupon
                    disponible.
                  </Text>
                </View>
              )}
              <View style={styles.containerBottom}>
                <View style={styles.ViewButton}>
                  <Pressable
                    style={[styles.button, styles.buttonOpen]}
                    onPress={() => addCoupon()}
                  >
                    <Text style={styles.textStyle}>Ajouter à votre liste</Text>
                  </Pressable>
                  <Pressable
                    style={[styles.button, styles.buttonClose]}
                    onPress={() => setStateModal(false)}
                  >
                    <Text style={styles.textStyle}>Fermer</Text>
                  </Pressable>
                </View>
              </View>
            </View>
          </View>
        </View>
      )}
    </Modal>
  );
};

// define your styles
const styles = StyleSheet.create({
  centeredView: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    marginTop: 22,
  },
  modalView: {
    position: "absolute",
    margin: 20,
    backgroundColor: "white",
    borderRadius: 20,
    padding: 35,
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
    /*width: 350,
    minHeight: 400,*/
  },
  titleText: {
    fontSize: 18,
    fontWeight: "bold",
    textAlign: "left",
    marginBottom: 10,
  },
  modalText: {
    marginBottom: 15,
  },

  /* Modal view's bottom */
  containerBottom: {
    marginTop: 20,
  },
  ViewButton: {
    flexDirection: "row",
    justifyContent: "space-between",
  },
  button: {
    borderRadius: 20,
    padding: 15,
    elevation: 2,
  },
  buttonOpen: {
    backgroundColor: "green",
  },
  buttonClose: {
    backgroundColor: "#FF5D3B",
  },
  textStyle: {
    color: "white",
    fontWeight: "bold",
    textAlign: "center",
  },
});

//make this component available to the app
export default ModalCoupon;
