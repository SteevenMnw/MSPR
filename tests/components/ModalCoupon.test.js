import React from "react";
import { render } from "@testing-library/react-native";

import ModalCoupon from "../../src/components/ModalCoupon";

it("renders", () => {
  render(<ModalCoupon visible={true} idCoupon={1} />);
});

it("modalShowTrue", () => {
  const { getByText, getByTestId, queryByText } = render(
    <ModalCoupon visible={true} idCoupon={4} />
  );

  expect(getByTestId("descriptionCoupon").props.childrens).toBe(
    "20% de r√©duction sur tous les produits du site."
  );
});
