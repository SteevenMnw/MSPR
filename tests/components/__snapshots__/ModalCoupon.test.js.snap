// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginTop": 22,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 20,
          "elevation": 5,
          "margin": 20,
          "padding": 35,
          "position": "absolute",
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 4,
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 10,
              "textAlign": "left",
            }
          }
        >
          Description du coupon : 
        </Text>
        <Text
          style={
            Object {
              "marginBottom": 15,
            }
          }
        />
        <View
          style={
            Object {
              "marginTop": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "borderRadius": 20,
                    "elevation": 2,
                    "padding": 15,
                  },
                  Object {
                    "backgroundColor": "green",
                  },
                ]
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                Ajouter Ã  votre liste
              </Text>
            </View>
            <View
              accessible={true}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "borderRadius": 20,
                    "elevation": 2,
                    "padding": 15,
                  },
                  Object {
                    "backgroundColor": "#FF5D3B",
                  },
                ]
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                Fermer
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
